connections/README.md

```diff
@@ -0,0 +1,325 @@
+new streams can be easily defined by libp2p applications.
+
+Streams are routed to application-defined handler functions based on their
+protocol id string. Incoming stream requests will propose a protocol id to use
+for the stream using `multistream-select`, and the peer accepting the stream
+request will determine if there are any registered handlers capable of handling
+the protocol. If no handlers are found, the peer will respond to the proposal
+with `"na"`.
+
+When registering protocol handlers, it's possible to use a custom predicate or
+"match function", which will receive incoming protocol ids and return `true` if
+the handler supports the protocol. This allows more flexible behavior than exact
+literal matching, which is the default behavior if no match function is
+provided.
+
+## Practical Considerations
+
+This section will go over a few aspects of connection establishment and state
+management that are worth considering when implementing libp2p.
+
+### Interoperability
+
+Support for connection security protocols and stream multiplexers varies across
+libp2p implementations. 
+
+To support the widest variety of peers, implementations should support a
+baseline "stack" of security and multiplexing protocols.
+
+At the time of writing, the recommended baseline security protocol is
+[SECIO][secio-spec], which is supported in all current libp2p implementations.
+
+Note that while SECIO is also the current default security protocol, that is
+likely to change with the further adoption of [TLS][tls-libp2p]. However,
+support for SECIO as a fallback will likely be recommended for some time after
+TLS becomes the default.
+
+The recommended baseline stream multiplexer is [mplex][mplex], which provides a
+very simple programmatic API and is supported in all current libp2p
+implementations.
+
+### State Management 
+
+While the connection establishment process itself does not require any
+persistent state, some state management is useful to assist bootstrapping and
+maintain resource limits.
+
+#### Peer Metadata Storage
+
+It's recommended that libp2p implementations provide a persistent metadata
+storage interface that contains at minimum the peer id and last known valid
+addresses for each peer. This allows you to more easily "catch
+back up" and rejoin a dense network between invocations of your libp2p
+application without having to rely on a few bootstrap nodes and random DHT walks
+to build up a routing table.
+
+Even during a single invocation of an application, you're likely to benefit from
+an in-memory metadata storage facility, which will allow you to cache addresses
+for connection resumption. Designing a storage interface which can be backed by
+memory or persistent storage will let you swap in whichever is appropriate for
+your use case and stage of development.
+
+For examples, see [go-libp2p-peerstore][go-libp2p-peerstore] and
+[js-peer-book][js-peer-book].
+
+#### Connection Limits
```

Comment:
Should we link this to the connection manager spec? I know it is WIP, but most of this information will also be there